service/storage_service.cc

```diff
@@ -840,9 +858,12 @@ future<> storage_service::handle_state_replacing(inet_address replacing_node) {
     }
     auto existing_node = *existing_node_opt;
     auto existing_tokens = get_tokens_for(existing_node);
-    auto replacing_tokens = get_tokens_for(replacing_node);
-    slogger.info("Node {} is replacing existing node {} with host_id={}, existing_tokens={}, replacing_tokens={}",
-            replacing_node, existing_node, host_id, existing_tokens, replacing_tokens);
+
+    auto existing_host_id = _gossiper.get_host_id(existing_node);
+    auto replacing_host_id = _gossiper.get_host_id(replacing_node);
+
+    slogger.info("Node {} is replacing existing node {}, existing_host_id={}, replacing_host_id={}, existing_tokens={}, replacing_tokens={},",
+            replacing_node, existing_node, existing_host_id, replacing_host_id, existing_tokens, replacing_tokens);
```

Comment:
I converting to printing nodes as "{}/{}", host_id, inet_address instead of separate "tagged" fields in the printout
for moving towards relying on host_id to identify the node rather than the inet address.

Also, shouldn't the existing tokens and replacing token be the same?
I think that part belongs to debug level, not info.

Reply:
It is for debug purpose. We can definitely tweak the log.

Reply:
We should check the tokens advertised by the replacing node and the node to be replaced are  indeed the same. Reject otherwise. A bogus node might send a wrong tokens.

Reply:
> We should check the tokens advertised by the replacing node and the node to be replaced are indeed the same. Reject otherwise. A bogus node might send a wrong tokens.

I agree.  May be this needs to be separated into a standalone patch

Reply:
How about extending https://github.com/scylladb/scylladb/blob/5ff4b8b5f83a040090e752b05ea352cd0604f876/service/storage_service.cc#L504-L514 to verify the bootstrap tokens?

Reply:
It is not about the check on the replacing node itself.  The replacing node will use the correct tokens (it learns the tokens of the node to be replaced from the existing cluster). 

The check is about when the existing node sees the replacing node becomes NORMAL, it has to verify the replacing node has used the exact the same token as the node to be replaced. Since the replacing node does not use the same host id, they only way the existing nodes know the new node is replacing is that the replacing node has used the same tokens belong to one of the node in the ring.

In theory, there are two things could go wrong:

1) The existing node uses wrong token for the node to be replaced but the replacing node uses correct one.

2) The existing node uses correct token for the node to be replaced but the replacing node uses the wrong one.

But in practice, the chance is pretty low.

Reply:
The existing node is invalidated as soon as the replace operation is committed.
What if it hibernates for longer and in between the cluster bootstrapped yet another node?
Then its owned tokens would not match any single node.
The whole method of mapping the tokens to ip addresses is writing on thin ice... If after the gossip shadow round it sees that no other node thinks it is a peer of it, based on its host_id (assuming the host_id is NOT migrated in replace), then it can assume it was replaced or removed and courteously bow and leave the stage.